File: autocontext/src/autocontext/integrations/browser/chrome_cdp.py
"""Thin CDP-backed browser session helpers."""

from __future__ import annotations

import json
from datetime import UTC, datetime
from typing import Any, Protocol
from uuid import uuid4

from autocontext.integrations.browser.contract.models import (
    BrowserAuditEvent,
    BrowserSessionConfig,
    BrowserSnapshot,
)
from autocontext.integrations.browser.contract.types import BrowserAction
from autocontext.integrations.browser.evidence import BrowserArtifactPaths, BrowserEvidenceStore
from autocontext.integrations.browser.policy import (
    BrowserPolicyDecision,
    evaluate_browser_action_policy,
)
from autocontext.integrations.browser.types import BrowserSessionPort
from autocontext.integrations.browser.validate import (
    validate_browser_action,
    validate_browser_audit_event,
    validate_browser_snapshot,
)

_SNAPSHOT_EXPRESSION = """
(() => {
  const candidates = Array.from(
    document.querySelectorAll("a,button,input,select,textarea,[role],[tabindex]")
  ).slice(0, 200);
  const refs = candidates.map((element, index) => ({
    id: `@e${index + 1}`,
    role: element.getAttribute("role") ?? element.tagName.toLowerCase(),
    name:
      element.getAttribute("aria-label") ??
      element.getAttribute("name") ??
      element.textContent?.trim() ??
      null,
    text: element.textContent?.trim() ?? null,
    selector: null,
    disabled: element.hasAttribute("disabled"),
  }));
  return {
    url: window.location.href,
    title: document.title ?? "",
    visibleText: document.body?.innerText ?? "",
    refs,
    html: document.documentElement?.outerHTML ?? "",
  };
})()
""".strip()


class ChromeCdpTransport(Protocol):
    async def send(self, method: str, params: dict[str, Any] | None = None) -> dict[str, Any]: ...
    async def close(self) -> None: ...


class ChromeCdpSession(BrowserSessionPort):
    """Policy-aware CDP session wrapper with local evidence capture."""

    def __init__(
        self,
        *,
        session_id: str,
        config: BrowserSessionConfig,
        transport: ChromeCdpTransport,
        evidence_store: BrowserEvidenceStore | None = None,
    ) -> None:
        self.session_id = session_id
        self.config = config
        self.transport = transport
        self.evidence_store = evidence_store
        self._current_url = "about:blank"
        self._domains_enabled = False
        self._ref_selectors: dict[str, str] = {}

    async def navigate(self, url: str) -> BrowserAuditEvent:
        action = self._build_action("navigate", {"url": url})
        decision = evaluate_browser_action_policy(self.config, action)
        if not decision.allowed:
            return self._record_action_result(
                action=action,
                decision=decision,
                before_url=self._current_url,
                after_url=self._current_url,
                message="navigation blocked by browser policy",
            )

        await self._ensure_domains_enabled()
        before_url = self._current_url
        await self.transport.send("Page.navigate", {"url": url})
        self._current_url = url
        return self._record_action_result(
            action=action,
            decision=decision,
            before_url=before_url,
            after_url=url,
            message="navigation allowed",
        )

    async def snapshot(self) -> BrowserSnapshot:
        action = self._build_action(
            "snapshot",
            {
                "captureHtml": True,
                "captureScreenshot": bool(self.config.captureScreenshots),
            },
        )
        await self._ensure_domains_enabled()

        response = await self.transport.send(
            "Runtime.evaluate",
            {
                "expression": _SNAPSHOT_EXPRESSION,
                "returnByValue": True,
                "awaitPromise": True,
            },
        )
        payload = _extract_result_value(response)
        refs = payload.get("refs")
        parsed_refs = refs if isinstance(refs, list) else []
        self._ref_selectors = {
            str(ref.get("id")): str(ref.get("selector"))
            for ref in parsed_refs
            if isinstance(ref, dict) and ref.get("id") and ref.get("selector")
        }

        screenshot_base64: str | None = None
        if bool(self.config.captureScreenshots):
            screenshot_response = await self.transport.send("Page.captureScreenshot", {"format": "png"})
            raw_data = screenshot_response.get("data")
            screenshot_base64 = str(raw_data) if isinstance(raw_data, str) else None

        artifacts = self._persist_snapshot_artifacts(
            basename=str(action.actionId),
            html=payload.get("html") if isinstance(payload.get("html"), str) else None,
            screenshot_base64=screenshot_base64,
        )

        url = payload.get("url")
        self._current_url = str(url) if isinstance(url, str) and url else self._current_url
        return validate_browser_snapshot({
            "schemaVersion": "1.0",
            "sessionId": self.session_id,
            "capturedAt": _utcnow(),
            "url": self._current_url,
            "title": str(payload.get("title") or ""),
            "refs": parsed_refs,
            "visibleText": str(payload.get("visibleText") or ""),
            "htmlPath": artifacts["htmlPath"],
            "screenshotPath": artifacts["screenshotPath"],
        })

    async def click(self, ref: str) -> BrowserAuditEvent:
        action = self._build_action("click", {"ref": ref})
        decision = evaluate_browser_action_policy(self.config, action)
        if not decision.allowed:
            return self._record_action_result(
                action=action,
                decision=decision,
                before_url=self._current_url,
                after_url=self._current_url,
            )

        await self._ensure_domains_enabled()
        selector = self._selector_for_ref(ref)
        await self.transport.send(
            "Runtime.evaluate",
            {
                "expression": _click_expression(selector),
                "returnByValue": True,
                "awaitPromise": True,
            },
        )
        return self._record_action_result(
            action=action,
            decision=decision,
            before_url=self._current_url,
            after_url=self._current_url,
            message="click allowed",
        )

    async def fill(
        self,
        ref: str,
        text: str,
        *,
        field_kind: str | None = None,
    ) -> BrowserAuditEvent:
        action = self._build_action("fill", {"ref": ref, "text": text, "fieldKind": field_kind})
        decision = evaluate_browser_action_policy(self.config, action)
        if not decision.allowed:
            return self._record_action_result(
                action=action,
                decision=decision,
                before_url=self._current_url,
                after_url=self._current_url,
                message="fill blocked by browser policy",
            )

        await self._ensure_domains_enabled()
        selector = self._selector_for_ref(ref)
        await self.transport.send(
            "Runtime.evaluate",
            {
                "expression": _fill_expression(selector, text),
                "returnByValue": True,
                "awaitPromise": True,
            },
        )
        return self._record_action_result(
            action=action,
            decision=decision,
            before_url=self._current_url,
            after_url=self._current_url,
            message="fill allowed",
        )

    async def press(self, key: str) -> BrowserAuditEvent:
        action = self._build_action("press", {"key": key})
        decision = evaluate_browser_action_policy(self.config, action)
        if not decision.allowed:
            return self._record_action_result(
                action=action,
                decision=decision,
                before_url=self._current_url,
                after_url=self._current_url,
            )

        await self._ensure_domains_enabled()
        await self.transport.send(
            "Runtime.evaluate",
            {
                "expression": _press_expression(key),
                "returnByValue": True,
                "awaitPromise": True,
            },
        )
        return self._record_action_result(
            action=action,
            decision=decision,
            before_url=self._current_url,
            after_url=self._current_url,
            message="key press allowed",
        )

    async def screenshot(self, name: str) -> BrowserAuditEvent:
        action = self._build_action("screenshot", {"name": name})
        decision = evaluate_browser_action_policy(self.config, action)
        if not decision.allowed:
            return self._record_action_result(
                action=action,
                decision=decision,
                before_url=self._current_url,
                after_url=self._current_url,
            )

        await self._ensure_domains_enabled()
        response = await self.transport.send("Page.captureScreenshot", {"format": "png"})
        screenshot_base64 = response.get("data")
        artifacts = self._persist_snapshot_artifacts(
            basename=name,
            screenshot_base64=str(screenshot_base64) if isinstance(screenshot_base64, str) else None,
        )
        return self._record_action_result(
            action=action,
            decision=decision,
            before_url=self._current_url,
            after_url=self._current_url,
            message="screenshot captured",
            artifacts=artifacts,
        )

    async def close(self) -> None:
        await self.transport.close()

    async def _ensure_domains_enabled(self) -> None:
        if self._domains_enabled:
            return
        await self.transport.send("Page.enable", {})
        await self.transport.send("Runtime.enable", {})
        self._domains_enabled = True

    def _build_action(self, action_type: str, params: dict[str, Any]) -> BrowserAction:
        return validate_browser_action({
            "schemaVersion": "1.0",
            "actionId": _new_id("act"),
            "sessionId": self.session_id,
            "timestamp": _utcnow(),
            "type": action_type,
            "params": params,
        })

    def _record_action_result(
        self,
        *,
        action: BrowserAction,
        decision: BrowserPolicyDecision,
        before_url: str | None,
        after_url: str | None,
        message: str | None = None,
        artifacts: BrowserArtifactPaths | None = None,
    ) -> BrowserAuditEvent:
        event = validate_browser_audit_event({
            "schemaVersion": "1.0",
            "eventId": _new_id("evt"),
            "sessionId": self.session_id,
            "actionId": str(action.actionId),
            "kind": "action_result",
            "allowed": decision.allowed,
            "policyReason": decision.reason,
            "timestamp": _utcnow(),
            "message": message,
            "beforeUrl": before_url,
            "afterUrl": after_url,
            "artifacts": artifacts or _empty_artifacts(),
        })
        if self.evidence_store is not None:
            self.evidence_store.append_audit_event(event)
        return event

    def _persist_snapshot_artifacts(
        self,
        *,
        basename: str,
        html: str | None = None,
        screenshot_base64: str | None = None,
    ) -> BrowserArtifactPaths:
        if self.evidence_store is None:
            return _empty_artifacts()
        return self.evidence_store.persist_snapshot_artifacts(
            session_id=self.session_id,
            basename=basename,
            html=html,
            screenshot_base64=screenshot_base64,
        )

    def _selector_for_ref(self, ref: str) -> str:
        return self._ref_selectors.get(ref, ref)


def _extract_result_value(response: dict[str, Any]) -> dict[str, Any]:
    result = response.get("result")
    if not isinstance(result, dict):
        return {}
    value = result.get("value")
    if not isinstance(value, dict):
        return {}
    return value


def _click_expression(selector: str) -> str:
    selector_json = json.dumps(selector)
    return f"""
(() => {{
  const element = document.querySelector({selector_json});
  if (!element) return {{ ok: false, error: "selector_not_found" }};
  element.click();
  return {{ ok: true }};
}})()
""".strip()


def _fill_expression(selector: str, text: str) -> str:
    selector_json = json.dumps(selector)
    text_json = json.dumps(text)
    return f"""
(() => {{
  const element = document.querySelector({selector_json});
  if (!element) return {{ ok: false, error: "selector_not_found" }};
  element.focus?.();
  if ("value" in element) {{
    element.value = {text_json};
  }}
  element.dispatchEvent(new Event("input", {{ bubbles: true }}));
  element.dispatchEvent(new Event("change", {{ bubbles: true }}));
  return {{ ok: true }};
}})()
""".strip()


def _press_expression(key: str) -> str:
    key_json = json.dumps(key)
    return f"""
(() => {{
  const target = document.activeElement ?? document.body;
  if (!target) return {{ ok: false, error: "missing_target" }};
  target.dispatchEvent(new KeyboardEvent("keydown", {{ key: {key_json}, bubbles: true }}));
  target.dispatchEvent(new KeyboardEvent("keyup", {{ key: {key_json}, bubbles: true }}));
  return {{ ok: true }};
}})()
""".strip()


def _empty_artifacts() -> BrowserArtifactPaths:
    return {"htmlPath": None, "screenshotPath": None, "downloadPath": None}


def _new_id(prefix: str) -> str:
    return f"{prefix}_{uuid4().hex}"


def _utcnow() -> datetime:
    return datetime.now(UTC)


__all__ = ["ChromeCdpSession", "ChromeCdpTransport"]

File: ts/src/integrations/browser/chrome-cdp.ts
import { randomUUID } from "node:crypto";
import type {
  BrowserAction,
  BrowserAuditEvent,
  BrowserFieldKind,
  BrowserPolicyDecision,
  BrowserSessionConfig,
  BrowserSnapshot,
  BrowserSnapshotRef,
  BrowserValidationResult,
} from "./contract/index.js";
import {
  BROWSER_CONTRACT_SCHEMA_VERSION,
  validateBrowserAction,
  validateBrowserAuditEvent,
  validateBrowserSnapshot,
} from "./contract/index.js";
import type { BrowserArtifactPaths } from "./evidence.js";
import { BrowserEvidenceStore } from "./evidence.js";
import { evaluateBrowserActionPolicy } from "./policy.js";
import type { BrowserSessionPort } from "./types.js";

const SNAPSHOT_EXPRESSION = `
(() => {
  const cssEscape = (value) =>
    globalThis.CSS?.escape
      ? CSS.escape(value)
      : String(value).replace(/[^a-zA-Z0-9_-]/g, "\\\\$&");
  const selectorFor = (element) => {
    if (element.id) return "#" + cssEscape(element.id);
    const parts = [];
    let current = element;
    while (current && current.nodeType === Node.ELEMENT_NODE && current !== document.documentElement) {
      const tag = current.tagName.toLowerCase();
      const parent = current.parentElement;
      if (!parent) {
        parts.unshift(tag);
        break;
      }
      const siblings = Array.from(parent.children).filter((sibling) => sibling.tagName === current.tagName);
      const index = siblings.indexOf(current) + 1;
      parts.unshift(siblings.length > 1 ? tag + ":nth-of-type(" + index + ")" : tag);
      current = parent;
      if (parts.length >= 4) break;
    }
    return parts.join(" > ");
  };
  const candidates = Array.from(
    document.querySelectorAll("a,button,input,select,textarea,[role],[tabindex]")
  ).slice(0, 200);
  const refs = candidates.map((element, index) => ({
    id: \`@e\${index + 1}\`,
    role: element.getAttribute("role") ?? element.tagName.toLowerCase(),
    name:
      element.getAttribute("aria-label") ??
      element.getAttribute("name") ??
      element.textContent?.trim() ??
      null,
    text: element.textContent?.trim() ?? null,
    selector: selectorFor(element),
    disabled: element.hasAttribute("disabled"),
  }));
  return {
    url: window.location.href,
    title: document.title ?? "",
    visibleText: document.body?.innerText ?? "",
    refs,
    html: document.documentElement?.outerHTML ?? "",
  };
})()
`.trim();

export interface ChromeCdpTransport {
  send(method: string, params?: Record<string, unknown>): Promise<Record<string, unknown>>;
  close(): Promise<void>;
}

export interface ChromeCdpSessionOpts {
  readonly sessionId: string;
  readonly config: BrowserSessionConfig;
  readonly transport: ChromeCdpTransport;
  readonly evidenceStore?: BrowserEvidenceStore;
}

type BrowserActionFor<TType extends BrowserAction["type"]> = Extract<BrowserAction, { type: TType }>;

export class ChromeCdpSession implements BrowserSessionPort {
  readonly config: BrowserSessionConfig;
  readonly sessionId: string;
  readonly transport: ChromeCdpTransport;
  readonly evidenceStore?: BrowserEvidenceStore;

  private currentUrl = "about:blank";
  private domainsEnabled = false;
  private readonly refSelectors = new Map<string, string>();

  constructor(opts: ChromeCdpSessionOpts) {
    this.sessionId = opts.sessionId;
    this.config = opts.config;
    this.transport = opts.transport;
    this.evidenceStore = opts.evidenceStore;
  }

  async navigate(url: string): Promise<BrowserAuditEvent> {
    const action = buildAction(this.sessionId, "navigate", { url });
    const decision = evaluateBrowserActionPolicy(this.config, action);
    if (!decision.allowed) {
      return this.recordActionResult({
        action,
        decision,
        beforeUrl: this.currentUrl,
        afterUrl: this.currentUrl,
        message: "navigation blocked by browser policy",
      });
    }

    await this.ensureDomainsEnabled();
    const beforeUrl = this.currentUrl;
    await this.transport.send("Page.navigate", { url });
    this.currentUrl = url;
    return this.recordActionResult({
      action,
      decision,
      beforeUrl,
      afterUrl: url,
      message: "navigation allowed",
    });
  }

  async snapshot(): Promise<BrowserSnapshot> {
    const action = buildAction(this.sessionId, "snapshot", {
      captureHtml: true,
      captureScreenshot: this.config.captureScreenshots,
    });
    await this.ensureDomainsEnabled();

    const response = await this.transport.send("Runtime.evaluate", {
      expression: SNAPSHOT_EXPRESSION,
      returnByValue: true,
      awaitPromise: true,
    });
    const payload = extractResultValue(response);
    const refs = extractRefs(payload.refs);

    this.refSelectors.clear();
    for (const ref of refs) {
      if (typeof ref.selector === "string" && ref.selector.length > 0) {
        this.refSelectors.set(ref.id, ref.selector);
      }
    }

    let screenshotBase64: string | null = null;
    if (this.config.captureScreenshots) {
      const screenshotResponse = await this.transport.send("Page.captureScreenshot", { format: "png" });
      screenshotBase64 = typeof screenshotResponse.data === "string" ? screenshotResponse.data : null;
    }

    const artifacts = this.persistSnapshotArtifacts({
      basename: action.actionId,
      html: typeof payload.html === "string" ? payload.html : null,
      screenshotBase64,
    });

    if (typeof payload.url === "string" && payload.url.length > 0) {
      this.currentUrl = payload.url;
    }

    return assertValidDocument(
      "browser snapshot",
      {
        schemaVersion: BROWSER_CONTRACT_SCHEMA_VERSION,
        sessionId: this.sessionId,
        capturedAt: new Date().toISOString(),
        url: this.currentUrl,
        title: typeof payload.title === "string" ? payload.title : "",
        refs,
        visibleText: typeof payload.visibleText === "string" ? payload.visibleText : "",
        htmlPath: artifacts.htmlPath,
        screenshotPath: artifacts.screenshotPath,
      },
      validateBrowserSnapshot,
    );
  }

  async click(ref: string): Promise<BrowserAuditEvent> {
    const action = buildAction(this.sessionId, "click", { ref });
    const decision = evaluateBrowserActionPolicy(this.config, action);
    if (!decision.allowed) {
      return this.recordActionResult({
        action,
        decision,
        beforeUrl: this.currentUrl,
        afterUrl: this.currentUrl,
      });
    }

    await this.ensureDomainsEnabled();
    const beforeUrl = this.currentUrl;
    const selector = this.selectorForRef(ref);
    await this.transport.send("Runtime.evaluate", {
      expression: buildClickExpression(selector),
      returnByValue: true,
      awaitPromise: true,
    });
    return this.recordInteractiveResult({
      action,
      decision,
      beforeUrl,
      message: "click allowed",
    });
  }

  async fill(
    ref: string,
    text: string,
    opts: { fieldKind?: BrowserFieldKind } = {},
  ): Promise<BrowserAuditEvent> {
    const action = buildAction(this.sessionId, "fill", {
      ref,
      text,
      fieldKind: opts.fieldKind,
    });
    const decision = evaluateBrowserActionPolicy(this.config, action);
    if (!decision.allowed) {
      return this.recordActionResult({
        action,
        decision,
        beforeUrl: this.currentUrl,
        afterUrl: this.currentUrl,
        message: "fill blocked by browser policy",
      });
    }

    await this.ensureDomainsEnabled();
    const beforeUrl = this.currentUrl;
    const selector = this.selectorForRef(ref);
    await this.transport.send("Runtime.evaluate", {
      expression: buildFillExpression(selector, text),
      returnByValue: true,
      awaitPromise: true,
    });
    return this.recordInteractiveResult({
      action,
      decision,
      beforeUrl,
      message: "fill allowed",
    });
  }

  async press(key: string): Promise<BrowserAuditEvent> {
    const action = buildAction(this.sessionId, "press", { key });
    const decision = evaluateBrowserActionPolicy(this.config, action);
    if (!decision.allowed) {
      return this.recordActionResult({
        action,
        decision,
        beforeUrl: this.currentUrl,
        afterUrl: this.currentUrl,
      });
    }

    await this.ensureDomainsEnabled();
    const beforeUrl = this.currentUrl;
    await this.transport.send("Runtime.evaluate", {
      expression: buildPressExpression(key),
      returnByValue: true,
      awaitPromise: true,
    });
    return this.recordInteractiveResult({
      action,
      decision,
      beforeUrl,
      message: "key press allowed",
    });
  }

  async screenshot(name: string): Promise<BrowserAuditEvent> {
    const action = buildAction(this.sessionId, "screenshot", { name });
    const decision = evaluateBrowserActionPolicy(this.config, action);
    if (!decision.allowed) {
      return this.recordActionResult({
        action,
        decision,
        beforeUrl: this.currentUrl,
        afterUrl: this.currentUrl,
      });
    }

    await this.ensureDomainsEnabled();
    const response = await this.transport.send("Page.captureScreenshot", { format: "png" });
    const artifacts = this.persistSnapshotArtifacts({
      basename: name,
      screenshotBase64: typeof response.data === "string" ? response.data : null,
    });
    return this.recordActionResult({
      action,
      decision,
      beforeUrl: this.currentUrl,
      afterUrl: this.currentUrl,
      message: "screenshot captured",
      artifacts,
    });
  }

  async close(): Promise<void> {
    await this.transport.close();
  }

  private async ensureDomainsEnabled(): Promise<void> {
    if (this.domainsEnabled) {
      return;
    }
    await this.transport.send("Page.enable", {});
    await this.transport.send("Runtime.enable", {});
    this.domainsEnabled = true;
  }

  private recordActionResult(opts: {
    readonly action: BrowserAction;
    readonly decision: BrowserPolicyDecision;
    readonly beforeUrl: string | null;
    readonly afterUrl: string | null;
    readonly message?: string;
    readonly artifacts?: BrowserArtifactPaths;
  }): BrowserAuditEvent {
    const rawEvent: BrowserAuditEvent = {
      schemaVersion: BROWSER_CONTRACT_SCHEMA_VERSION,
      eventId: newId("evt"),
      sessionId: this.sessionId,
      actionId: opts.action.actionId,
      kind: "action_result",
      allowed: opts.decision.allowed,
      policyReason: opts.decision.reason,
      timestamp: new Date().toISOString(),
      message: opts.message ?? null,
      beforeUrl: opts.beforeUrl,
      afterUrl: opts.afterUrl,
      artifacts: opts.artifacts ?? emptyArtifacts(),
    };
    const event = assertValidDocument("browser audit event", rawEvent, validateBrowserAuditEvent);
    this.evidenceStore?.appendAuditEvent(event);
    return event;
  }

  private persistSnapshotArtifacts(opts: {
    readonly basename: string;
    readonly html?: string | null;
    readonly screenshotBase64?: string | null;
  }): BrowserArtifactPaths {
    return (
      this.evidenceStore?.persistSnapshotArtifacts({
        sessionId: this.sessionId,
        basename: opts.basename,
        html: opts.html,
        screenshotBase64: opts.screenshotBase64,
      }) ?? emptyArtifacts()
    );
  }

  private selectorForRef(ref: string): string {
    return this.refSelectors.get(ref) ?? ref;
  }

  private async recordInteractiveResult(opts: {
    readonly action: BrowserAction;
    readonly decision: BrowserPolicyDecision;
    readonly beforeUrl: string | null;
    readonly message: string;
  }): Promise<BrowserAuditEvent> {
    const afterUrl = await this.readCurrentUrl();
    this.currentUrl = afterUrl;
    const afterDecision = evaluateNavigationUrlPolicy(this.config, afterUrl);
    if (!afterDecision.allowed) {
      return this.recordActionResult({
        action: opts.action,
        decision: afterDecision,
        beforeUrl: opts.beforeUrl,
        afterUrl,
        message: "interaction navigated outside browser policy",
      });
    }
    return this.recordActionResult({
      action: opts.action,
      decision: opts.decision,
      beforeUrl: opts.beforeUrl,
      afterUrl,
      message: opts.message,
    });
  }

  private async readCurrentUrl(): Promise<string> {
    const response = await this.transport.send("Runtime.evaluate", {
      expression: "(() => window.location.href)()",
      returnByValue: true,
      awaitPromise: true,
    });
    const result = response.result;
    if (!isRecord(result)) {
      return this.currentUrl;
    }
    const value = result.value;
    return typeof value === "string" && value.length > 0 ? value : this.currentUrl;
  }
}

function buildAction<TType extends BrowserAction["type"]>(
  sessionId: string,
  type: TType,
  params: BrowserActionFor<TType>["params"],
): BrowserActionFor<TType> {
  const rawAction = {
    schemaVersion: BROWSER_CONTRACT_SCHEMA_VERSION,
    actionId: newId("act"),
    sessionId,
    timestamp: new Date().toISOString(),
    type,
    params,
  } as unknown as BrowserActionFor<TType>;
  return assertValidDocument(
    "browser action",
    rawAction,
    validateBrowserAction,
  );
}

function assertValidDocument<T>(
  label: string,
  value: T,
  validate: (input: unknown) => BrowserValidationResult,
): T {
  const result = validate(value);
  if (!result.valid) {
    throw new TypeError(`invalid ${label}: ${result.errors.join("; ")}`);
  }
  return value;
}

function extractResultValue(response: Record<string, unknown>): Record<string, unknown> {
  const result = response.result;
  if (!isRecord(result)) {
    return {};
  }
  const value = result.value;
  return isRecord(value) ? value : {};
}

function extractRefs(raw: unknown): BrowserSnapshotRef[] {
  if (!Array.isArray(raw)) {
    return [];
  }
  return raw.flatMap((entry) => {
    if (!isRecord(entry) || typeof entry.id !== "string") {
      return [];
    }
    return [
      {
        id: entry.id,
        role: typeof entry.role === "string" ? entry.role : undefined,
        name: typeof entry.name === "string" ? entry.name : undefined,
        text: typeof entry.text === "string" ? entry.text : undefined,
        selector: typeof entry.selector === "string" ? entry.selector : undefined,
        disabled: typeof entry.disabled === "boolean" ? entry.disabled : undefined,
      },
    ];
  });
}

function buildClickExpression(selector: string): string {
  return `
(() => {
  const element = document.querySelector(${JSON.stringify(selector)});
  if (!element) return { ok: false, error: "selector_not_found" };
  element.click();
  return { ok: true };
})()
`.trim();
}

function buildFillExpression(selector: string, text: string): string {
  return `
(() => {
  const element = document.querySelector(${JSON.stringify(selector)});
  if (!element) return { ok: false, error: "selector_not_found" };
  element.focus?.();
  if ("value" in element) {
    element.value = ${JSON.stringify(text)};
  }
  element.dispatchEvent(new Event("input", { bubbles: true }));
  element.dispatchEvent(new Event("change", { bubbles: true }));
  return { ok: true };
})()
`.trim();
}

function buildPressExpression(key: string): string {
  return `
(() => {
  const target = document.activeElement ?? document.body;
  if (!target) return { ok: false, error: "missing_target" };
  target.dispatchEvent(new KeyboardEvent("keydown", { key: ${JSON.stringify(key)}, bubbles: true }));
  target.dispatchEvent(new KeyboardEvent("keyup", { key: ${JSON.stringify(key)}, bubbles: true }));
  return { ok: true };
})()
`.trim();
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function emptyArtifacts(): BrowserArtifactPaths {
  return {
    htmlPath: null,
    screenshotPath: null,
    downloadPath: null,
  };
}

function evaluateNavigationUrlPolicy(
  config: BrowserSessionConfig,
  url: string,
): BrowserPolicyDecision {
  return evaluateBrowserActionPolicy(config, {
    schemaVersion: BROWSER_CONTRACT_SCHEMA_VERSION,
    actionId: "act_interaction_url_probe",
    sessionId: "session_interaction_url_probe",
    timestamp: new Date().toISOString(),
    type: "navigate",
    params: { url },
  });
}

function newId(prefix: string): string {
  return `${prefix}_${randomUUID().replaceAll("-", "")}`;
}
